### ISSUE TITLE
`trainer.evaluate` always creates a new MLFlow run, separate from the one used during `train()`

### ISSUE THREAD

**Issue by nathan-az** [huggingface/transformers#35074]
### System Info

`transformers` version: 4.46.1
`mlflow==2.18.0`

I have confirmed this on my local machine (mac) and our training cluster (8x H100 using standard nvcr image), using a local and external MLFlow tracking server, so this is not likely to be very environment dependent outside the above.

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```py
from time import sleep

from transformers import Trainer, TrainingArguments, AutoModelForSequenceClassification, AutoTokenizer
from datasets import load_dataset
import mlflow
import os


with mlflow.start_run() as run:
    run_id = run.info.run_id
os.environ["MLFLOW_RUN_ID"] = run_id
print("Original run id:", run_id)

# Load a pre-trained model and tokenizer
model_name = "distilbert-base-uncased"
model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=2)
tokenizer = AutoTokenizer.from_pretrained(model_name)

# Load a small dataset
dataset = load_dataset("imdb", split="train[:100]")

# Tokenize the dataset
def tokenize(batch):
    return tokenizer(batch['text'], padding=True, truncation=True, max_length=128)

tokenized_dataset = dataset.map(tokenize, batched=True)

# Define training arguments
training_args = TrainingArguments(
    output_dir="./results",
    evaluation_strategy="epoch",
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    num_train_epochs=1,
    logging_steps=1,
    report_to="mlflow"
)

# Initialize Trainer
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_dataset,
    eval_dataset=tokenized_dataset,
)

# Train the model
trainer.train()

# Evaluate the model
eval_results = trainer.evaluate()
print(eval_results)

sleep(10)
runs_dir = "./mlruns/0"
runs = [x for x in os.listdir(runs_dir) if x != "meta.yaml"]
for run_id in runs:
    print(f"Metric files for {run_id}")
    metrics_path = os.path.join(runs_dir, run_id, "metrics")
    print(*os.listdir(metrics_path), sep="\n")
    print("="*16)
```
Running the above script should confirm that two separate runs are created, and that metrics files are placed in each.

### Expected behavior

I expected that all logging would be pushed to the existing run ID (unless there is a reason that this is not the case).

Apologies if this is documented behaviour, but I could not find a reason.

**Comment by Rocketknight1** [comment]
cc @muellerzr @SunMarc

**Comment by LysandreJik** [comment]
@nathan-az do you have an idea of the fix? If so, feel free to open a PR with the fix, we're missing some bandwith in the new year. Thanks 🤗

**Comment by nathan-az** [comment]
@LysandreJik I'm starting to think having a separate run is actually a good thing, for a couple of reasons:
* The trainer can store the same metrics in time series during `.train`, and we don't want to overwrite these during `.evaluate`
* I often use one tracking run for multiple stages of model training (to co-locate artifacts)

A middleground her might be for `evaluate` to still create a new run, but for it to be a child run if an existing `MLFLOW_RUN_ID` is set in the environment, at least improving organisation. I'm happy to take a look, but won't get around to it for a few weeks. I'll link this issue in the eventual draft PR.

**Comment by SunMarc** [comment]
As Lysandre is off for a couple of weeks, I assigned you to the issue ! Excited to see the draft PR

**Comment by nathan-az** [comment]
Apologies @SunMarc - I haven't found the time to get to this and am unlikely to any time soon. I'd encourage anybody interested to take a look into this!

As mentioned there might have to be a bit of discussion around the best course of action, given that the user likely doesn't want to overwrite the metrics written during training. Two options I can think of:
* Use the existing run but prefix the metric names to make them unique, or
* Create a child run if an existing run ID is found

**Comment by gitedmond** [comment]
Hi @SunMarc, I’d be interested in taking this on if it’s still available. I saw that @nathan-az previously encouraged anyone interested to pick it up, although the issue still appears assigned.

The nested-run approach seems like the cleanest behavior to me. It would keep evaluation metrics separate from the training metric history while still associating them with the same experiment, and it would scale more clearly if evaluate() is called multiple times or against different datasets.

My proposed scope would be:

* evaluation performed during train() continues logging to the active training run
* a standalone evaluate() after training creates an evaluation child run when MLFLOW_RUN_ID identifies the parent
* behavior remains unchanged when no parent run ID is provided
* regression tests verify the parent-child relationship and repeated evaluation behavior

Would this align with the preferred design, and would it be okay for me to take the issue?

**Comment by YeonwooSung** [comment]
@gitedmond @SunMarc checking in before anyone opens a PR.

The original path is still there: `on_train_end` closes the run, then a later standalone `evaluate()` hits `on_log` and lands in a new MLflow run. No open PR. @nathan-az already stepped back (2025-03-13).

@gitedmond you offered this on 2026-08-29 with a nested child run and asked for maintainer approval. Are you still on it? If yes, I'll leave it with you.

If not, @SunMarc could you pick a default first? Changing it is an observability contract change, and Trackio's recent re-init after train end (#46935) is a different contract.

Options:
- A. child run under the train parent for standalone `evaluate()` (gitedmond's suggestion)
- B. keep the same run open after `train()` (what the original report expected)
- C. same run + a metric prefix
- D. keep today's sibling-run behavior; opt-in only (`MLFLOW_NESTED_RUN` already exists)

**Comment by gitedmond** [comment]
@YeonwooSung Thanks for checking in. I am still interested in working on this. I’m leaning toward option A, but I’ll wait for @SunMarc’s guidance on the preferred behavior before opening the PR.
